Cover it everything (#5374)
cpinitiative/usaco-guide: solutions/silver/cf-1176E.mdx

---
id: cf-1176E
source: CF
title: Cover It!
author: David Zhou
---

[Official Analysis (C++)](https://codeforces.com/blog/entry/67598)

## Explanation

To solve this problem, we need to find a bipartite coloring where every node is either selected or adjacent to one that is.

> Perform [BFS](https://usaco.guide/silver/graph-traversal?lang=cpp) starting from any node and calculate the distance (level) from the root to each node. Every node will have a parity of even or odd.

Reviewer: ```suggestion
Perform [BFS](/silver/graph-traversal) starting from any node and calculate the distance (level) from the root to each node. Every node will have a parity of even or odd.
```

> Since BFS visits every node at the same level before moving on, edges must connect nodes of opposite parities. This provides the bipartition that separates every node.

Reviewer: idk if it would be good to frame it this way, but at least to me the most intuitive way to understand this problem is that we just need to do a bipartite coloring of a spanning tree, and print the smaller set of nodes

(which obv works by the definition of the problem, since adding non-tree edges can only 'improve' our state)

> The categories are disjoint and exhaustive, so one group must have $\leq n/2$ nodes. We can output the smaller of the two sets to satisfy the problem's requirements.

Reviewer: ```suggestion
The categories are disjoint and exhaustive, so one group must have less than or equal to $\left \lfloor \frac{n}{2} \right \rfloor$ nodes. We can output the smaller of the two sets to satisfy the problem's requirements.
```

## Implementation

**Time Complexity:** $\mathcal{O}(N+M)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {

> 	int t;
> 	cin >> t;
> 	while (t--) {

Reviewer: ```suggestion
	int test_num;
	cin >> test_num;
	for (int t = 0; t < test_num; t++) {
```

just a bit of a thing we do

		int n, m;
		cin >> n >> m;
		vector<vector<int>> adj(n);
		for (int i = 0; i < m; i++) {
			int u, v;
			cin >> u >> v;
			u--;
			v--;
			adj[u].push_back(v);
			adj[v].push_back(u);
		}

		// BFS from node 0
		queue<int> q;
		q.push(0);
		vector<int> dist(n, -1);
		dist[0] = 0;
		while (!q.empty()) {
			int curr = q.front();
			q.pop();
			for (int next : adj[curr]) {
				if (dist[next] == -1) {
					dist[next] = dist[curr] + 1;
					q.push(next);
				}
			}
		}

		// separates nodes by parity and outputs smaller group
		vector<int> even, odd;
		for (int i = 0; i < n; i++) {

> 			dist[i] % 2 == 0 ? even.push_back(i + 1) : odd.push_back(i + 1);

Reviewer: if you write it with a ternary i would write it like

`(dist[i] % 2 == 0 ? even : odd).push_back(i + 1);`

		}
		if (even.size() < odd.size()) {
			cout << even.size() << endl;
			for (int num : even) { cout << num << " "; }
		} else {
			cout << odd.size() << endl;
			for (int num : odd) { cout << num << " "; }
		}
		cout << endl;
	}
}
```

</CPPSection>

<JavaSection>

```java
import java.io.*;
import java.util.*;

public class Cover {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		PrintWriter pw = new PrintWriter(new OutputStreamWriter(System.out));

		int t = Integer.parseInt(br.readLine());

> 		while (t-- > 0) {

Reviewer: same thing with test_num instead of t with cpp

			String[] nm = br.readLine().split(" ");
			int n = Integer.parseInt(nm[0]);
			int m = Integer.parseInt(nm[1]);

> 			List<List<Integer>> adj = new ArrayList<>();

Reviewer: `List<Integer>[]` is better imo
